### PR TITLE
fix: replace sync.Once with retryable OAuth endpoint discovery

### PR DESCRIPTION
**Pull request by GowthamShanmugam**
sync.Once caused permanent auth failures when the kube-apiserver was briefly unavailable during discovery. Now discoverOpenShiftOAuth is called on-demand (matching openshift/oauth-proxy behavior), allowing natural retry on the next login attempt once the API server recovers.

Resolves: [RHOAIENG-67051](https://redhat.atlassian.net/browse/RHOAIENG-67051)

## Checklist:

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [ ] I have created a feature (non-master) branch for my PR.
- [ ] I have written tests for my code changes.

## Summary by CodeRabbit

* **Refactor**
  * Switched OpenShift OAuth discovery to on-demand, mutex-guarded discovery to avoid persistent cached state and reduce memory footprint.

* **Tests**
  * Added retry tests covering temporary discovery failures and recovery.
  * Expanded coverage for login URL resolution and token redemption under transient discovery outages.

### CODE REVIEW

File: providers/openshift.go [hunk 1]
```diff
@@ -131,27 +126,22 @@ func (p *OpenShiftProvider) GetLoginURL(redirectURI, state, _ string, extraParam
...
+			logger.Errorf("Failed to discover OpenShift OAuth endpoints: %v", err)
 			logger.Printf("Please configure --login-url and --redeem-url manually")
 			return ""
 		}
-		p.LoginURL = p.discoveredLogin
+		p.LoginURL = loginURL
 		if p.RedeemURL == nil || p.RedeemURL.String() == "" {
-			p.RedeemURL = p.discoveredRedeem
+			p.RedeemURL = redeemURL
 		}
```
[the remark this answers is not in the document]
Comment (GowthamShanmugam) [reply providers/openshift.go]: check now

File: providers/openshift_test.go [hunk 2]
```diff
@@ -412,6 +413,174 @@ func TestOpenShiftProviderCreateSessionFromTokenWithGroups(t *testing.T) {
...
+	serverURL, err := url.Parse(server.URL)
+	require.NoError(t, err)
+	t.Setenv("KUBERNETES_SERVICE_HOST", serverURL.Hostname())
+	t.Setenv("KUBERNETES_SERVICE_PORT", serverURL.Port())
+
+	// newOpenShiftClient checks DefaultTransport for InsecureSkipVerify
+	originalTransport := http.DefaultTransport
+	http.DefaultTransport = &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true}, // #nosec G402
+	}
```
[the remark this answers is not in the document]
Comment (GowthamShanmugam) [reply providers/openshift_test.go]: fixed

File: providers/openshift_test.go [hunk 3]
```diff
@@ -412,6 +414,183 @@ func TestOpenShiftProviderCreateSessionFromTokenWithGroups(t *testing.T) {
 	assert.Equal(t, 3, len(session.Groups))
 }
 
+func TestOpenShiftProviderDiscoveryRetriesAfterFailure(t *testing.T) {
```
Comment (ugiordan) [inline providers/openshift_test.go:417]: Consider adding a few edge-case tests that would strengthen confidence in the change:

 1. **Concurrent access under `-race`**: the `discoveryMu` is the key addition here but nothing exercises it under contention. A test that spawns multiple goroutines calling `GetLoginURL` simultaneously would validate it:

  ```go
  func TestOpenShiftProviderGetLoginURLConcurrentDiscovery(t *testing.T) {
      server := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
          if r.URL.Path == "/.well-known/oauth-authorization-server" {
              time.Sleep(50 * time.Millisecond) // simulate latency
              w.Header().Set("Content-Type", "application/json")
              w.WriteHeader(http.StatusOK)
              w.Write([]byte(`{
                  "authorization_endpoint": "https://oauth.example.com/authorize",
                  "token_endpoint": "https://oauth.example.com/token"
              }`))
              return
          }
          w.WriteHeader(http.StatusNotFound)
      }))
      defer server.Close()

      serverURL, _ := url.Parse(server.URL)
      t.Setenv("KUBERNETES_SERVICE_HOST", serverURL.Hostname())
      t.Setenv("KUBERNETES_SERVICE_PORT", serverURL.Port())

      originalTransport := http.DefaultTransport
      http.DefaultTransport = &http.Transport{
          TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
      }
      defer func() { http.DefaultTransport = originalTransport }()

      provider := &OpenShiftProvider{ProviderData: &ProviderData{}}

      var wg sync.WaitGroup
      for i := 0; i < 10; i++ {
          wg.Add(1)
          go func() {
              defer wg.Done()
              result := provider.GetLoginURL("http://localhost/callback", "state", "", url.Values{})
              assert.NotEmpty(t, result)
          }()
      }
      wg.Wait()

      assert.Equal(t, "https://oauth.example.com/authorize", provider.LoginURL.String())
      assert.Equal(t, "https://oauth.example.com/token", provider.RedeemURL.String())
  }
```
  Run with go test -race ./providers/ to verify no data races.

2. **Persistent failure**:

  ```go
  func TestOpenShiftProviderDiscoveryPersistentFailureThenRecovery(t *testing.T) {
      callCount := 0
      server := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
          if r.URL.Path == "/.well-known/oauth-authorization-server" {
              callCount++
              if callCount <= 3 {
                  w.WriteHeader(http.StatusServiceUnavailable)
                  w.Write([]byte("API server unavailable"))
                  return
              }
              w.Header().Set("Content-Type", "application/json")
              w.WriteHeader(http.StatusOK)
              w.Write([]byte(`{
                  "authorization_endpoint": "https://oauth.example.com/authorize",
                  "token_endpoint": "https://oauth.example.com/token"
              }`))
              return
          }
          w.WriteHeader(http.StatusNotFound)
      }))
      defer server.Close()

      serverURL, _ := url.Parse(server.URL)
      t.Setenv("KUBERNETES_SERVICE_HOST", serverURL.Hostname())
      t.Setenv("KUBERNETES_SERVICE_PORT", serverURL.Port())

      originalTransport := http.DefaultTransport
      http.DefaultTransport = &http.Transport{
          TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
      }
      defer func() { http.DefaultTransport = originalTransport }()

      provider := &OpenShiftProvider{ProviderData: &ProviderData{}}

      // Calls 1-3: persistent failure, each should return empty without corrupting state
      for i := 0; i < 3; i++ {
          result := provider.GetLoginURL("http://localhost/callback", "state", "", url.Values{})
          assert.Empty(t, result, "GetLoginURL should return empty on failure %d", i+1)
          assert.Nil(t, provider.LoginURL, "LoginURL should remain nil after failure %d", i+1)
          assert.Nil(t, provider.RedeemURL, "RedeemURL should remain nil after failure %d", i+1)
      }

      // Call 4: server recovers, discovery should succeed
      result := provider.GetLoginURL("http://localhost/callback", "state", "", url.Values{})
      assert.NotEmpty(t, result, "GetLoginURL should succeed once server recovers")
      assert.Equal(t, "https://oauth.example.com/authorize", provider.LoginURL.String())
      assert.Equal(t, "https://oauth.example.com/token", provider.RedeemURL.String())
      assert.Equal(t, 4, callCount, "Should have attempted discovery exactly 4 times")
  }
```

  3. GetLoginURL discovers, Redeem reuses:
```
  func TestOpenShiftProviderRedeemReusesDiscoveredURL(t *testing.T) {
      discoveryCount := 0
      server := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
          switch r.URL.Path {
          case "/.well-known/oauth-authorization-server":
              discoveryCount++
              w.Header().Set("Content-Type", "application/json")
              w.WriteHeader(http.StatusOK)
              w.Write([]byte(`{
                  "authorization_endpoint": "` + server.URL + `/oauth/authorize",
                  "token_endpoint": "` + server.URL + `/oauth/token"
              }`))
          case "/oauth/token":
              w.Header().Set("Content-Type", "application/json")
              w.WriteHeader(http.StatusOK)
              w.Write([]byte(`{"access_token": "test-token"}`))
          default:
              w.WriteHeader(http.StatusNotFound)
          }
      }))
      defer server.Close()

      serverURL, _ := url.Parse(server.URL)
      t.Setenv("KUBERNETES_SERVICE_HOST", serverURL.Hostname())
      t.Setenv("KUBERNETES_SERVICE_PORT", serverURL.Port())

      originalTransport := http.DefaultTransport
      http.DefaultTransport = &http.Transport{
          TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
      }
      defer func() { http.DefaultTransport = originalTransport }()

      provider := &OpenShiftProvider{
          ProviderData: &ProviderData{
              ClientID:     "test-client",
              ClientSecret: "test-secret",
          },
      }

      // GetLoginURL triggers discovery, sets both LoginURL and RedeemURL
      result := provider.GetLoginURL("http://localhost/callback", "state", "", url.Values{})
      assert.NotEmpty(t, result)
      assert.Equal(t, 1, discoveryCount, "GetLoginURL should trigger exactly one discovery call")

      // Redeem should use the already-discovered RedeemURL, no new discovery
      session, err := provider.Redeem(context.Background(), "http://localhost/callback", "code123", "")
      assert.NoError(t, err)
      assert.NotNil(t, session)
      assert.Equal(t, "test-token", session.AccessToken)
      assert.Equal(t, 1, discoveryCount, "Redeem should not trigger additional discovery when RedeemURL is already set")
  }
```

**Review by ugiordan — APPROVED** [review]

**Review by ugiordan — APPROVED** [review]

File: providers/openshift_test.go (hunk 3 above)
Comment (GowthamShanmugam) [reply providers/openshift_test.go:417]: I will add this in new PR
